src/connectors/impls/gbq/writer/sink.rs

```diff
@@ -0,0 +1,660 @@
+                    val.as_bytes().ok_or_else(|| {
+                        ErrorKind::BigQueryTypeMismatch("bytes", val.value_type())
+                    })?,
+                ),
+                result,
+            );
+        }
+        TableType::Json => {
+            warn!("Found a field of type JSON, this is not supported, ignoring.");
+        }
+        TableType::Interval => {
+            warn!("Found a field of type Interval, this is not supported, ignoring.");
+        }
+
+        TableType::Unspecified => {
+            warn!("Found a field of unspecified type - ignoring.");
+        }
+    }
+
+    Ok(())
+}
+
+impl JsonToProtobufMapping {
+    pub fn new(vec: &Vec<TableFieldSchema>) -> Self {
+        let descriptor = map_field("table", vec);
+
+        Self {
+            descriptor: descriptor.0,
+            fields: descriptor.1,
+        }
+    }
+
+    pub fn map(&self, value: &Value) -> Result<Vec<u8>> {
+        let mut result = vec![];
+        if let Some(obj) = value.as_object() {
+            for (key, val) in obj {
+                if let Some(field) = self.fields.get(&key.to_string()) {
+                    encode_field(val, field, &mut result)?;
+                }
+            }
+        }
+
+        Ok(result)
+    }
+
+    pub fn descriptor(&self) -> &DescriptorProto {
+        &self.descriptor
+    }
+}
+impl GbqSink {
+    pub async fn new(config: Config) -> Result<Self> {
+        Ok(Self {
+            client: None,
+            write_stream: None,
+            mapping: None,
+            config,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl Sink for GbqSink {
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        event: Event,
+        ctx: &SinkContext,
+        _serializer: &mut EventSerializer,
+        _start: u64,
+    ) -> Result<SinkReply> {
+        let client = self
+            .client
+            .as_mut()
+            .ok_or(ErrorKind::BigQueryClientNotAvailable(
+                "The client is not connected",
+            ))?;
+        let write_stream =
+            self.write_stream
+                .as_ref()
+                .ok_or(ErrorKind::BigQueryClientNotAvailable(
+                    "The write stream is not available",
+                ))?;
+        let mapping = self
+            .mapping
+            .as_ref()
+            .ok_or(ErrorKind::BigQueryClientNotAvailable(
+                "The mapping is not available",
+            ))?;
+
+        let mut serialized_rows = vec![];
+
+        for data in event.value_iter() {
+            serialized_rows.push(mapping.map(data)?);
+        }
+
+        let request = AppendRowsRequest {
+            write_stream: write_stream.name.clone(),
+            offset: None,
+            trace_id: "".to_string(),
+            rows: Some(append_rows_request::Rows::ProtoRows(ProtoData {
+                writer_schema: Some(ProtoSchema {
+                    proto_descriptor: Some(mapping.descriptor().clone()),
```

Comment:
this only needs to be provided upon the first request (after connect). We could store the full `ProtoSchema` in an `Option<ProtoSchema>` and extract it with a `self.proto_schema.take()` here and have it be a `None` all the time, saving 1 clone. :)
We could do the same with `trace_id` if we would have it configured on the connector.

Reply:
interesting, very interesting... :oldmanshakesfistatcloud:

Reply:
The underlying client is `tonic`, which is pretty good about handling long-lived connections, as far as I know. On the other hand - we're using HTTP here, so I expect the connection might not survive forever.

Reply:
Oh if it's HTTP not HTTP2 then that explains it, http will be counted as 1 connection per request, using http2/quick would probably solve that

Reply:
We should be able to use grpc status codes for this purpose - perhaps thats something worth investigating and optimising in a separate ticket ( this would impact otel similarly and in future other grpc endpoints )